### DISCUSSION TITLE
.chezmoi.hostname source priority

### DISCUSSION THREAD

**Discussion by stinovlas** [twpayne/chezmoi#3649] [General]
Hi! I use `.chezmoi.hostname` variable in my scripts to differentiate between individual machines. I bumped into a weird situation. In one of my machine, I use `/etc/hosts` to point different services to `127.0.0.1`, for development reasons. This totally wrecks the value of `.chezmoi.hostname`. It doesn't matter what `/etc/hostname` contains or what `hostname` command returns. It just uses the first non-local domain from `/etc/hosts` and considers that to be a hostname.

I find this really confusing. Why did you decide to prioritize `/etc/hosts` parsing before `/etc/hostname`? And is there any way to override this auto-detected value?

**Comment by twpayne** [comment]
What is the contents of your `/etc/hosts` file?

**Reply by stinovlas** [reply]
Here's an anonymized example:

```
# Host addresses
127.0.0.1  localhost
127.0.1.1  alfheim
::1        localhost ip6-localhost ip6-loopback
ff02::1    ip6-allnodes
ff02::2    ip6-allrouters

127.0.0.1  alfheim.local  # <-- I added this to fix the chezmoi autodetection!
127.0.0.1  project1.my.work.domain.cz
127.0.0.1  project2.my.work.domain.cz
127.0.0.1  project3.my.work.domain.cz
127.0.0.1  project4.my.work.domain.cz
127.0.0.1  project5.my.work.domain.cz
```
Before I added the line indicated above, `chezmoi` autodetected the hostname as `projec1`.

I could get rid of this configuration by using custom local DNS resolver, but it's additional setup and nothing ever collided with my `/etc/hosts` until now. And the added line, although not pretty, does also provide a workaround for my problem.

I'm more interested in the origin of this domain decision. Why do you consider `/etc/hosts` before `/etc/hostname`? Or do you disregard `/etc/hostname` because it only contains hostname with one label (i.e. `alfheim`)? It's quite common to have just one-label hostnames on development machines that don't serve as public servers.

**Reply by Lockszmith-GH** [reply]
@stinovlas [This](https://github.com/twpayne/chezmoi/blob/88b586dd4f6f9c61413631369f6a7195a6935c2f/internal/chezmoi/chezmoi.go#L132) is where hostname resolution takes place.

I'm guessing the reasoning has to do with trial and error with different OSs and their detection options.

What system are you running?

As for overridding - you can just poplate a variable of your own by calling `hostname`, and use that instead. But let Tom take a look first.

**Reply by twpayne** [reply]
chezmoi is looking for a fully-qualified domain name associated with a loopback address. Before you added `alfheim.local`, the first FQDN with a loopback address in your `/etc/hosts` is `project1.my.work.domain.cz`. chezmoi is working as intended here as your `/etc/hosts` clearly states that `project1.my.work.domain.cz` has IP address 127.0.0.1, i.e. is the localhost.

The fix for you is to include your actual FQDN earlier in the file, as you did when you added `alfheim.local`.

chezmoi's hostname detection has evolved to be robust across multiple operating systems and configurations. It cannot, however, detect misconfigurations.

If you have to edit `/etc/hosts` to test your software, I strongly recommend exploring the topic of service discovery.

**Reply by stinovlas** [reply]
I wouldn't say my system is misconfigured. When I run `hostnamectl hostname`, I get correct value, `alfheim`. The same value is returned by `hostname` command and also stored in `/etc/hostname`. For all intents and purposes, the hostname of my machine is `alfheim`. But `chezmoi` tries to parse my `/etc/hosts` instead and makes assumptions based on its content. This is highly non-standard for a linux app.

I'm not saying that adding additional hostnames to `/etc/hosts` is a good practice, but that's not the point. Of course, I could add the correct hostname to `chezmoi.toml` and use it instead of `.chezmoi.hostname`, but that seems like unnecessary duplication.

But after all, there are some ways around this and `chezmoi` is still awesome regardless. This is just one thing that's bitten me and I found it really confusing. Maybe `chezmoi` could call `hostname --fqdn` which returns hostname in full FQDN format (or some Go equivalent)? That would fix the hostname resolution issue quite nicely (because the whole problems stems from the problem that `chezmoi` expects `.` in the hostname and `os.Hostname()` does not return FQDN).

**Reply by halostatue** [reply]
chezmoi uses several different steps in hostname resolution, as I mentioned in a [comment on #3655](https://github.com/twpayne/chezmoi/issues/3655#issuecomment-2010587529). We also explicitly *do not* call `hostname` or `hostnamectl` (mentioned in a [comment on #3407](https://github.com/twpayne/chezmoi/issues/3407#issuecomment-1854440752)).

Both of those comments have links to other discussions on issues and PRs.

**Reply by twpayne** [reply]
Note that you can use `hostnamectl` with chezmoi. Put the following in your config file template:

```
[data]
    hostname = {{ output "hostnamectl" "hostname" | trim | quote }}
```

and then use `{{ .hostname }}` in your templates.

**Reply by stinovlas** [reply]
This is a good suggestion and it occurred to me as well. This is probably the way to go.

**Reply by felipecrs** [reply]
I'm using:

```
# .chezmoi.yaml.tmpl

{{- $fqdnHostname := .chezmoi.fqdnHostname -}}
{{- if and (not (contains "." $fqdnHostname)) (lookPath "hostname") -}}
{{-   $fqdnHostname = output "hostname" "-f" | trim -}}
{{-   if not (contains "." $fqdnHostname) -}}
{{-     $fqdnHostname = output "sh" "-c" "hostname -A 2>/dev/null | cut -d' ' -f1" | trim -}}
{{-     if not (contains "." $fqdnHostname) -}}
{{-       $fqdnHostname = .chezmoi.fqdnHostname -}}
{{-       writeToStdout (printf "⚠️ WARNING: could not determine the fqdn hostname, falling back to '%s'\n" $fqdnHostname) -}}
{{-     end -}}
{{-   end -}}
{{- end -}}

data:
    fqdn_hostname: "{{ $fqdnHostname }}"
```

Which works better on my environments.

- Refs https://github.com/felipecrs/dotfiles/pull/121
